Source/Request.swift

```diff
@@ -342,7 +346,9 @@ open class Request {
     }
 
     /// Finishes this `Request` and starts the response serializers.
-    func finish() {
+    func finish(error: Error? = nil) {
+        if let error = error { self.error = error }
```

Comment:
@jshier I'd love your feedback here. Since the `RequestRetrier` can now modify the error or pass the original through, we need to set the error back on the `Request`. However, the API doesn't have any way to know where the retrier actually changed the error or just passed the original through. This is a limitation of the `Result` type since I cannot customize the `Error` type to be an optional `Error`. This requires the `RequestRetrier` to either forward the original error through a `.failure`, or create a new error altogether.

Rather than introduce another entire API, I decided to just extend the `finish` API to support an optional `Error` type. Super open to feedback here as well as the change to have the retrier take `Result` type in the `completion`. It **does** allow us to send custom errors back in the retrier, but forwarding the original error and not being able to tell if you changed the error in the `Session` seem like design flaws. Should we abandon `Result` in the retrier case so that we can solve these two issues?

Reply:
The has been redesigned in b9d0b5d7. The `finish` API is still the same as before, but it will now get an `AFError.requestRetryFailed` error with both the retry error and origin error.